### PR TITLE
Submission feature.

### PR DESCRIPTION
**Pull request by isht3**
This PR covers 

- `evalai challenge CHALLENGE phase PHASE submit FILE`

- `evalai submission SUBMISSION`

### CODE REVIEW

**Comment by RishabhJain2018** [comment]
@isht3 More Conflicts :D

**Comment by isht3** [comment]
@RishabhJain2018 OMG. On it! :joy:

**Comment by isht3** [comment]
@RishabhJain2018 Done!

**Comment by RishabhJain2018** [comment]
@isht3 How do I invoke the second command since I don't have submission Id?

File: evalai/utils/submissions.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+from evalai.utils.auth import get_request_header
+from evalai.utils.config import API_HOST_URL, EVALAI_ERROR_CODES
+from evalai.utils.urls import URLS
+from evalai.utils.common import validate_token
+
+
+def submit_a_file(challenge_id, phase_id, file):
+    """
+    Function submit a file to a challenge
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `Function to submit a file to a challenge`

File: evalai/utils/submissions.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+from click import echo, style
+
+from evalai.utils.auth import get_request_header
+from evalai.utils.config import API_HOST_URL, EVALAI_ERROR_CODES
+from evalai.utils.urls import URLS
+from evalai.utils.common import validate_token
+
+
+def submit_a_file(challenge_id, phase_id, file):
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `submit_a_file` -> `make_submission`

File: evalai/utils/submissions.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    response = response.json()
+    echo(style("\nYour file {} was successfully submitted.\n".format(file.name),
+               fg="green", bold=True))
+
+
+def pretty_print_submission_details(submission):
+    """
+    Function to print details of a submission
+    """
+    team_title = "\n{}".format(style(submission['participant_team_name'],
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `team_name`

File: evalai/utils/submissions.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                                     bold=True, fg="green"))
+    sid = "Submission ID: {}\n".format(style(str(submission['id']),
+                                       bold=True, fg="blue"))
+    team = "{} {}".format(team_title, sid)
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `team_name`

File: evalai/utils/submissions.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    status = style("\nSubmission Status : {}\n".format(submission['status']), bold=True)
+    execution_time = style("\nExecution Time : {}\n".format(submission['execution_time']), bold=True)
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `style("\nExecution Time (sec) : {}\n".format(submission['execution_time']), bold=True)`

File: evalai/utils/urls.py [hunk 6]
```diff
@@ -13,3 +13,5 @@ class URLS(Enum):
     participate_in_a_challenge = "/api/challenges/challenge/{}/participant_team/{}"
     challenge_phase_list = "/api/challenges/challenge/{}/challenge_phase"
     challenge_phase_detail = "/api/challenges/challenge/{}/challenge_phase/{}"
+    submit_a_file = "/api/jobs/challenge/{}/challenge_phase/{}/submission/"
```
Comment (RishabhJain2018) [inline evalai/utils/urls.py]: `make_submission`

File: evalai/utils/urls.py [hunk 7]
```diff
@@ -13,3 +13,5 @@ class URLS(Enum):
...
+    submission = "/api/jobs/submission/{}"
```
Comment (RishabhJain2018) [inline evalai/utils/urls.py]: `get_submission`

File: tests/data/submission_response.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+submission_result = """
+                {
+                    "challenge_phase": 7,
+                    "created_by": 4,
+                    "execution_time": "None",
+                    "id": 9,
+                    "input_file": "http://localhost:8000/media/submission_files/submission_9/2224fb89-6828-\
```
Comment (RishabhJain2018) [inline tests/data/submission_response.py]: `http://localhost:8000/` -> `http://testserver/`

File: tests/test_requests.py [hunk 9]
```diff
@@ -142,6 +149,26 @@ def test_participate_in_a_challenge_for_http_error_404(self):
...
+    @responses.activate
+    def test_display_submission_details_for_http_error_404(self):
+        runner = CliRunner()
+        result = runner.invoke(submission, ['9'])
+        response = result.output.rstrip()
+        url = "{}{}".format(API_HOST_URL, URLS.submission.value).format("9")
+        assert response == self.expected.format(url)
+
+    @responses.activate
+    def test_submit_a_file_for_http_error_404(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: Please change the name accordingly.

File: tests/test_requests.py [hunk 10]
```diff
@@ -398,3 +432,21 @@ def test_display_challenge_phase_detail_for_request_exception(self):
...
+
+    @responses.activate
+    def test_display_submission_details_for_request_exception(self):
+        runner = CliRunner()
+        result = runner.invoke(submission, ['9'])
+        response = result.output.strip()
+        assert response == "RequestException"
+
+    @responses.activate
+    def test_submit_a_file_for_request_exception(self):
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: same as above

File: tests/test_submissions.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+from evalai.challenges import challenge
+from evalai.submissions import submission
+from tests.data import submission_response
+
+from evalai.utils.challenges import API_HOST_URL
+from evalai.utils.urls import URLS
+from .base import BaseTestClass
+
+
+class TestSubmission(BaseTestClass):
```
Comment (RishabhJain2018) [inline tests/test_submissions.py]: `TestGetSubmissionDetails`

File: tests/test_submissions.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    def test_display_submission_details_with_no_argument(self):
+        expected = ("Usage: submission [OPTIONS] SUBMISSION COMMAND [ARGS]...\n"
+                    "\nError: Missing argument \"SUBMISSION\".\n")
+        runner = CliRunner()
+        result = runner.invoke(submission)
+        response = result.output
+        assert response == expected
+
+
+class TestSubmitAFile(BaseTestClass):
```
Comment (RishabhJain2018) [inline tests/test_submissions.py]: `TestMakeSubmission`

**Comment by isht3** [comment]
@RishabhJain2018

**Comment by RishabhJain2018** [comment]
@isht3 Please also return the command `evalai submission SUBMISSION` after the submission response that you are returning saying like `you can use evalai submission SUBMISSION to view your submission status` Do you see what I mean here?

File: evalai/utils/submissions.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            echo(style("Error: {}".format(response.json()["error"]), fg="red", bold=True))
+        else:
+            echo(err)
+        sys.exit(1)
+    except requests.exceptions.RequestException as err:
+        echo(err)
+        sys.exit(1)
+    response = response.json()
+    response_message = ("\nYour file {} with the ID {} was successfully submitted.\n\n"
+                        "You can use `evalai submission {}` to view this submission's status.\n").format(file.name,
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: Please change its colour
`You can use evalai submission {} to view above submission's status.` to white.

File: evalai/utils/submissions.py (hunk 3 above)
Comment (RishabhJain2018) [reply evalai/utils/submissions.py]: I think you missed this one!

File: evalai/utils/submissions.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                                       bold=True, fg="blue"))
+    team_name = "{} {}".format(team_title, sid)
+
+    status = style("\nSubmission Status : {}\n".format(submission['status']), bold=True)
+    execution_time = style("\nExecution Time (sec) : {}\n".format(submission['execution_time']), bold=True)
+
+    date = datetime.strptime(submission['submitted_at'], "%Y-%m-%dT%H:%M:%S.%fZ")
+    from_zone = tz.tzutc()
+    to_zone = tz.tzlocal()
+
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: please add comments here.

File: tests/test_submissions.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    def test_make_submission_when_file_is_not_valid(self):
+        expected = ("Usage: challenge phase submit [OPTIONS] FILE\n"
+                    "\nError: Invalid value for \"FILE\": Could not open file: file: No such file or directory\n")
+        runner = CliRunner()
+        result = runner.invoke(challenge, ['1', 'phase', '2', 'submit', 'file'])
+        response = result.output
+        assert response == expected
+
+    @responses.activate
+    def test_make_submission_is_valid(self):
```
Comment (RishabhJain2018) [inline tests/test_submissions.py]: `test_make_submission_when_file_is_valid`

**Comment by isht3** [comment]
@RishabhJain2018 Solved!

**Comment by RishabhJain2018** [comment]
@isht3 One more major thing that I missed is that user should have options to fill Submission Metadata like `method name`, `method description`, `project url`, `publication url` at the time of submission.
Let me know if you want to add it in this PR only or work on it as a different issue.

Rest looks good to me 🎉

File: evalai/utils/submissions.py [hunk 16]
```diff
@@ -7,12 +7,50 @@
...
-def pretty_print_my_submissions_data(submissions):
+def make_submission(challenge_id, phase_id, file):
+    """
+    Function to submit a file to a challenge
+    """
+    url = "{}{}".format(API_HOST_URL, URLS.make_submission.value)
+    url = url.format(challenge_id, phase_id)
+
+    headers = get_request_header()
+    file_data = {'input_file': file}
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `input_file`

File: evalai/utils/submissions.py [hunk 17]
```diff
@@ -7,12 +7,50 @@
...
+    data = {
+            'status': 'submitting',
+           }
+
+    try:
+        response = requests.post(
+                                url,
+                                headers=headers,
+                                files=file_data,
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: `input_file`

File: evalai/utils/submissions.py [hunk 18]
```diff
@@ -7,12 +7,50 @@
...
+            validate_token(response.json())
+            echo(style("Error: {}".format(response.json()["error"]), fg="red", bold=True))
+        else:
+            echo(err)
+        sys.exit(1)
+    except requests.exceptions.RequestException as err:
+        echo(err)
+        sys.exit(1)
+    response = response.json()
+    echo(style("\nYour file {} with the ID {} was successfully submitted.\n".format(file.name, response["id"]),
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: was -> is

**Comment by isht3** [comment]
@RishabhJain2018 I'll solve it in a new issue since I think works standalone. #67

**Comment by RishabhJain2018** [comment]
sounds good! Approving this one!
@varunagrawal Can you please review this once.

File: evalai/utils/submissions.py [hunk 19]
```diff
@@ -44,11 +44,13 @@ def make_submission(challenge_id, phase_id, file):
...
     response = response.json()
-    echo(style("\nYour file {} with the ID {} was successfully submitted.\n\n".format(file.name, response["id"]),
+    echo(style("\nYour file {} with the ID {} is successfully submitted.\n".format(file.name, response["id"]),
                fg="green", bold=True))
+    echo(style("You can use `evalai submission {}` to view this submission's status.\n".format(response["id"]),
+               bold=True))
 
 
-def pretty_print_my_submissions_data(submissions):
+def pretty_print_make_submission_data(submissions):
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: I think you mistakenly changed this line here!

File: evalai/utils/submissions.py [hunk 20]
```diff
@@ -106,25 +108,30 @@ def display_my_submission_details(challenge_id, phase_id):
     response = response.json()
 
     submissions = response["results"]
-    pretty_print_my_submissions_data(submissions)
+    pretty_print_make_submission_data(submissions)
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: same as above!

File: evalai/utils/submissions.py (hunk 14 above)
Comment (RishabhJain2018) [reply evalai/utils/submissions.py]: Ahh..you missed this one!

File: tests/test_challenges.py [hunk 21]
```diff
@@ -309,7 +309,7 @@ def setup(self):
         json_data = json.loads(submission_response.submission)
 
         url = "{}{}"
-        responses.add(responses.GET, url.format(API_HOST_URL, URLS.my_submissions.value).format("3", "7"),
+        responses.add(responses.GET, url.format(API_HOST_URL, URLS.make_submission.value).format("3", "7"),
```
Comment (RishabhJain2018) [inline tests/test_challenges.py]: please check this once!

File: tests/test_challenges.py [hunk 22]
```diff
@@ -369,7 +369,7 @@ def setup(self):
         json_data = json.loads(data)
 
         url = "{}{}"
-        responses.add(responses.GET, url.format(API_HOST_URL, URLS.my_submissions.value).format("3", "7"),
+        responses.add(responses.GET, url.format(API_HOST_URL, URLS.make_submission.value).format("3", "7"),
```
Comment (RishabhJain2018) [inline tests/test_challenges.py]: same as above!

File: tests/test_requests.py [hunk 23]
```diff
@@ -57,7 +57,7 @@ def setup(self):
                       status=404)
 
         # Submission URLS
-        responses.add(responses.GET, url.format(API_HOST_URL, URLS.my_submissions.value).format("3", "7"), status=404)
+        responses.add(responses.GET, url.format(API_HOST_URL, URLS.make_submission.value).format("3", "7"), status=404)
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: same as above!

File: tests/test_requests.py [hunk 24]
```diff
@@ -156,7 +156,7 @@ def test_display_my_submission_details_for_http_error_404(self):
         runner = CliRunner()
         result = runner.invoke(challenge, ['3', 'phase', '7', 'submissions'])
         response = result.output
-        url = "{}{}".format(API_HOST_URL, URLS.my_submissions.value).format("3", "7")
+        url = "{}{}".format(API_HOST_URL, URLS.make_submission.value).format("3", "7")
         expected = "{}{}".format(self.expected.format(url), "\n")
         assert response == expected
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: same as above!

File: tests/test_requests.py [hunk 25]
```diff
@@ -366,9 +366,8 @@ def setup(self):
                       body=Exception('...'))
 
         # Submission URLS
-        responses.add(responses.GET, url.format(API_HOST_URL, URLS.my_submissions.value).format("3", "7"),
+        responses.add(responses.GET, url.format(API_HOST_URL, URLS.make_submission.value).format("3", "7"),
```
Comment (RishabhJain2018) [inline tests/test_requests.py]: same as above!

File: evalai/utils/submissions.py [hunk 26]
```diff
@@ -44,11 +44,13 @@ def make_submission(challenge_id, phase_id, file):
         echo(err)
         sys.exit(1)
     response = response.json()
-    echo(style("\nYour file {} with the ID {} was successfully submitted.\n\n".format(file.name, response["id"]),
+    echo(style("\nYour file {} with the ID {} is successfully submitted.\n".format(file.name, response["id"]),
                fg="green", bold=True))
+    echo(style("You can use `evalai submission {}` to view this submission's status.\n".format(response["id"]),
+               bold=True))
```
Comment (RishabhJain2018) [inline evalai/utils/submissions.py]: Please revert all the changes below! I have requested only the above changes.

File: evalai/submissions.py [hunk 27]
```diff
@@ -1,9 +1,15 @@
...
+@click.group(invoke_without_command=True)
+@click.argument('SUBMISSION', type=int)
+def submission(submission):
+    """
+    View submission specific details.
+    """
+    """
+    Invoked by `evalai submission SUBMISSION`.
+    """
+    display_submission_details(submission)
```
Comment (varunagrawal) [inline evalai/submissions.py]: Is it `print_x` everywhere else? Or `display_x`? If it is print, please rename this to ensure consistency.

File: evalai/utils/submissions.py [hunk 28]
```diff
@@ -71,3 +109,54 @@ def display_my_submission_details(challenge_id, phase_id):
...
+    status = style("\nSubmission Status : {}\n".format(submission['status']), bold=True)
+    execution_time = style("\nExecution Time (sec) : {}\n".format(submission['execution_time']), bold=True)
+
+    # Converting UTC to local timezone
+    date = datetime.strptime(submission['submitted_at'], "%Y-%m-%dT%H:%M:%S.%fZ")
+    from_zone = tz.tzutc()
+    to_zone = tz.tzlocal()
+
+    date = date.replace(tzinfo=from_zone)
+    converted_date = date.astimezone(to_zone)
```
Comment (varunagrawal) [inline evalai/utils/submissions.py]: possibly put this in a separate `utils` module? I expect one to use it for a lot of similar things.

File: evalai/submissions.py (hunk 27 above)
Comment (isht3) [reply evalai/submissions.py]: It's `display_x` for fetching the data to be printed, `pretty_print_x` does the printing

File: evalai/utils/common.py [hunk 29]
```diff
@@ -34,3 +35,25 @@ def validate_token(response):
...
             echo(style("\nSorry, the token has expired. Please generate it again.\n", bold=True, bg="red"))
             sys.exit(1)
+
+
+def validate_date_format(date):
+    for fmt in ("%Y-%m-%dT%H:%M:%S.%fZ", "%Y-%m-%dT%H:%M:%SZ"):
+        try:
+            return datetime.strptime(date, fmt)
+        except ValueError:
+            pass
```
Comment (RishabhJain2018) [inline evalai/utils/common.py]: I think we should also an error when an exception occurs `raise ValueError('Invalid date format. Please check again!')`

File: evalai/utils/common.py [hunk 30]
```diff
@@ -34,3 +35,25 @@ def validate_token(response):
...
+    raise ValueError('no valid date format found')
```
Comment (RishabhJain2018) [inline evalai/utils/common.py]: `'Invalid date format. Please check again!'`

File: evalai/utils/common.py [hunk 31]
```diff
@@ -34,3 +35,25 @@ def validate_token(response):
         if (response['detail'] == 'Token has expired'):
             echo(style("\nSorry, the token has expired. Please generate it again.\n", bold=True, bg="red"))
             sys.exit(1)
+
+
+def validate_date_format(date):
+    for fmt in ("%Y-%m-%dT%H:%M:%S.%fZ", "%Y-%m-%dT%H:%M:%SZ"):
```
Comment (RishabhJain2018) [inline evalai/utils/common.py]: fmt -> `format`

File: evalai/utils/common.py (hunk 29 above)
Comment (isht3) [reply evalai/utils/common.py]: No, so the loop would terminate and it would only check the first format

**Pull request merged**